src/wallet/wallet.cpp

```diff
@@ -245,7 +245,7 @@ bool CWallet::Unlock(const SecureString& strWalletPassphrase, bool anonymizeOnly
 {
     SecureString strWalletPassphraseFinal;
 
-    if (!IsLocked())
+    if (!IsLocked() && !fWalletUnlockAnonymizeOnly) // was already fully unlocked, not only for anonymization
```

Comment:
It looks to me that the only difference between an unlocked wallet and a wallet unlocked for anonymization only is the value of the flag fWalletUnlockAnonymizeOnly.  They're both really unlocked and code that uses the wallet is just supposed to check that flag before allowing transactions to be sent (which is only the case for gui code as far as I can tell).